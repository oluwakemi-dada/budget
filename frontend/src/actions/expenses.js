import { v4 as uuid } from 'uuid';
import database from '../firebase/firebase';
import {
  ADD_EXPENSE,
  REMOVE_EXPENSE,
  EDIT_EXPENSE,
} from '../constants/expensesConstants';

// ADD_EXPENSE
export const addExpense = (expense) => (dispatch) => {
  database
    .ref('expenses')
    .push(expense)
    .then((ref) => {
      // dispatch add expense
      dispatch({
        type: ADD_EXPENSE,
        payload: { id: ref.key, ...expense },
      });
    });
};

// REMOVE_EXPENSE
export const removeExpense = (id) => ({
  type: REMOVE_EXPENSE,
  payload: id,
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: EDIT_EXPENSE,
  payload: {
    id,
    updates,
  },
});
