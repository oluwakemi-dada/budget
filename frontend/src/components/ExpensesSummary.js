import React from 'react';
import { useSelector } from 'react-redux';
import numeral from 'numeral';
import visibleExpenses from '../selectors/expenses';
import visibleExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = () => {
  // SELECTOR
  const expenses = useSelector((state) => state.expenses);
  const filters = useSelector((state) => state.filters);

  const visibleExp = visibleExpenses(expenses, filters);

  // VALUES
  const expenseCount = visibleExp.length;
  const expensesTotal = visibleExpensesTotal(visibleExp);

  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const expensesTotalNumeral = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totaling {expensesTotalNumeral}
      </h1>
    </div>
  );
};

export default ExpensesSummary;
