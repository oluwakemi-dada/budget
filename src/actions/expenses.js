import { v4 as uuid } from 'uuid';
import {
  ADD_EXPENSE,
  REMOVE_EXPENSE,
  EDIT_EXPENSE,
} from '../constants/expensesConstants';

// ADD_EXPENSE
export const addExpense = ({
  description,
  amount,
  createdAt,
  note = '',
} = {}) => ({
  type: ADD_EXPENSE,
  payload: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt,
  },
});

// REMOVE_EXPENSE
export const removeExpense = (id) => ({
  type: REMOVE_EXPENSE,
  payload: id,
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: EDIT_EXPENSE,
  payload: {
    id,
    updates,
  },
});
