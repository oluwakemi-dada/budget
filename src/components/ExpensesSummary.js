import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import visibleExpenses from '../selectors/expenses';
import visibleExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const expensesTotalNumeral = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totaling {expensesTotalNumeral}
      </h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const getVisibleExpenses = visibleExpenses(state.expenses, state.filters);
  return {
    expenseCount: getVisibleExpenses.length,
    expensesTotal: visibleExpensesTotal(getVisibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
