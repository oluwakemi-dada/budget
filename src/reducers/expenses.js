import {
  ADD_EXPENSE,
  REMOVE_EXPENSE,
  EDIT_EXPENSE,
} from '../constants/expensesConstants';

// Expenses Reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_EXPENSE:
      return [...state, action.payload];

    case REMOVE_EXPENSE:
      return state.filter(({ id }) => id !== action.payload);

    case EDIT_EXPENSE:
      return state.map((expense) => {
        console.log(action.payload.id);
        console.log(action.payload.updates);
        if (expense.id === action.payload.id) {
          return {
            ...expense,
            ...action.payload.updates,
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
