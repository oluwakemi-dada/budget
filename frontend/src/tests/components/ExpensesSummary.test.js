import React from 'react';
import { shallow } from 'enzyme';
import * as reactRedux from 'react-redux';
import ExpensesSummary from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

const filters = {
  text: '',
  sortBy: 'date',
  startDate: undefined,
  endDate: undefined,
};

describe('ExpensesSummary with no expense', () => {
  beforeEach(() => {
    useSelectorMock.mockImplementation((selector) => selector(mockStore));
  });

  afterEach(() => {
    useSelectorMock.mockClear();
  });

  const useSelectorMock = reactRedux.useSelector;

  const mockStore = {
    expenses: [],
    filters,
  };

  test('should return 0 if no expense', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('ExpensesSummary with a single expense', () => {
  beforeEach(() => {
    useSelectorMock.mockImplementation((selector) => selector(mockStore));
  });

  afterEach(() => {
    useSelectorMock.mockClear();
  });

  const useSelectorMock = reactRedux.useSelector;

  const mockStore = {
    expenses: [expenses[2]],
    filters,
  };

  test('should correctly add up a single expense', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('ExpensesSummary with multiple expenses', () => {
  beforeEach(() => {
    useSelectorMock.mockImplementation((selector) => selector(mockStore));
  });

  afterEach(() => {
    useSelectorMock.mockClear();
  });

  const useSelectorMock = reactRedux.useSelector;

  const mockStore = {
    expenses,
    filters,
  };

  test('should correctly add up multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
  });
});
