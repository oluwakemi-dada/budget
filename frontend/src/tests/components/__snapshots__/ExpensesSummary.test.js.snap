// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpensesSummary with a single expense should correctly add up a single expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $45.00
  </h1>
</div>
`;

exports[`ExpensesSummary with multiple expenses should correctly add up multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totaling 
    $1,141.95
  </h1>
</div>
`;

exports[`ExpensesSummary with no expense should return 0 if no expense 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totaling 
    $0.00
  </h1>
</div>
`;
